### DISCUSSION TITLE
Representing Loops within Egg

### DISCUSSION THREAD

**Discussion by Kixiron** [egraphs-good/egg#106] [Q&A] [answered]
I'm really struggling to figure out how to express more advanced control flow with egg, things like `for` and `while` loops. From exploring the problem space I feel like more traditional compiler constructs like lowering to tail-controlled `do ... while ...` loops and some sort of phi node are the most promising candidates, but I'm still not entirely sure how to do this effectively.
Modeling loops in a form that egg understands is something critical to whole-program optimization, autovectorization and other fancy types of optimization and I'm eager to get it working, thanks!

**Comment by Bastacyclop** [comment]
The original paper on equality saturation (https://www.cs.cornell.edu/~ross/publications/eqsat/) features loops, I would assume that you can re-use the same loop representation technique with egg?

**Reply by Kixiron** [reply]
It looks like the paper makes a distinction between phi and theta nodes, but I'm not sure what the meaningful distinction actually is
To quote them in section 2.1,
> We use θ nodes to represent values that vary inside of a loop. [...] As in gated SSA, our φ nodes are executable: the first (left-most) argument to φ is a selector that is used to select between the second and the third argument. [...] From a more formal point of view, each θ node produces a sequence of values, one value for each iteration of the loop. The first argument of a θ node is the value for the first iteration, whereas the second argument is a sequence that represents the values for the remaining iterations.

So if I'm understanding correctly, they're treating φ as a normal conditional branch node (`(if ?cond ?then ?else)`) while θ is doing the actual legwork of being what would *traditionally* be called a phi node in SSA IR, akin to LLVM's phi nodes

**Reply by mwillsey** [reply]
I think your intuition is right here. Phi/theta are pretty closely related, the difference being whether you view it as a sequence or not (and therefore need to use eval/pass nodes). The 'Related Work' section of the paper is helpful in understanding that these views are all variations on the same theme.

**Reply by Kixiron** [reply]
By the way, if you haven't seen 'em before I think you may be really interested in the [RVSDG](https://www.sjalander.com/research/pdf/sjalander-tecs2020.pdf), it's a pretty natural extension of equisat's capabilities and opens some *incredibly* interesting possibilities

**Reply by Kixiron** [reply]
I finally got around to implementing this but I've run into the roadblock of being unable to actually create phi nodes since they ultimately need to reference an eclass that uses them and egg doesn't have a way to create any kind of "delayed" eclass or enode
Example:
```rust
let x = egraph.add(Lang::Var("x"));

// This (rightfully) won't compile since `x_minus_one` doesn't exist yet
let phi = egraph.add(Lang::Phi([x, x_minus_one]));

// But we can't move `x_minus_one` before `phi` is created since it depends on `phi`
let one = egraph.add(Lang::Int(1));
let x_minus_one = egraph.add(Lang::Sub(phi, 1));
```

**Reply by Kixiron** [reply]
It also appears that egg panics when trying to extract a phi expression (eclass 12 is a node that encompasses eclass 2 (the phi node))
```
thread 'rewrites::tests::for_i_in_ten' panicked at 'Failed to extract from eclass 12', C:\Users\Chase\.cargo\registry\src\github.com-1ecc6299db9ec823\egg-0.6.0\src\extract.rs:206:21
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633\/library\std\src\panicking.rs:515
   1: std::panicking::begin_panic_fmt
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633\/library\std\src\panicking.rs:457
   2: egg::extract::Extractor<graph_asm::cost::OperatorCost, enum$<graph_asm::operator::Operator>, graph_asm::OperatorAnalyzer>::find_best_rec<graph_asm::cost::OperatorCost,enum$<graph_asm::operator::Operator>,graph_asm::OperatorAnalyzer>
             at C:\Users\Chase\.cargo\registry\src\github.com-1ecc6299db9ec823\egg-0.6.0\src\extract.rs:206
   3: egg::extract::Extractor<graph_asm::cost::OperatorCost, enum$<graph_asm::operator::Operator>, graph_asm::OperatorAnalyzer>::find_best<graph_asm::cost::OperatorCost,enum$<graph_asm::operator::Operator>,graph_asm::OperatorAnalyzer>
             at C:\Users\Chase\.cargo\registry\src\github.com-1ecc6299db9ec823\egg-0.6.0\src\extract.rs:198
   4: graph_asm::rewrites::tests::for_i_in_ten
             at .\src\rewrites\tests.rs:193
   5: graph_asm::rewrites::tests::for_i_in_ten::{{closure}}
             at .\src\rewrites\tests.rs:156
   6: core::ops::function::FnOnce::call_once<closure-0,tuple<>>
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633\library\core\src\ops\function.rs:227
   7: core::ops::function::FnOnce::call_once
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633\library\core\src\ops\function.rs:227
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
test rewrites::tests::for_i_in_ten ... FAILED
```
```rust
EClass {
    id: 2,
    nodes: [
        Symbol(
            "TOMBSTONE",
        ),
        Other(
            "phi",
            [
                0,
                3,
            ],
        ),
    ],
    data: OperatorData {
        constant: None,
        ty: None,
    },
    parents: [
        (
            Sub(
                [
                    2,
                    1,
                ],
            ),
            3,
        ),
        (
            Eq(
                [
                    2,
                    5,
                ],
            ),
            6,
        ),
        (
            Add(
                [
                    2,
                    13,
                ],
            ),
            14,
        ),
        (
            Add(
                [
                    13,
                    2,
                ],
            ),
            15,
        ),
    ],
}
```

**Reply by Bastacyclop** [reply]
I've played around with the ability to create empty e-classes in the past:
https://github.com/Bastacyclop/egg/blob/748d8ae2fe829db815f75d10b4c88b4a6fb2caaa/src/egraph.rs#L200-L212

It seemed to work fine, although I think I had to tweak some code that asssumed non-empty e-classes.

**Reply by mwillsey** [reply]
If you want to create cycles, you can first insert an acyclic term and then create a cycle with a union. So if you insert `f(a)` and then union that with `a`, you'll get a cycle. In your case, you'll just end up adding a "dummy" node.

Extraction will never extract a term with a cycle in it, so I think that's the issue.

**Reply by Kixiron** [reply]
Yep, that's what the backtrack refers to (I should have clarified that more): I added a tombstone enode into the egraph as a placeholder, unioned it with the final node and then removed the tombstone from the egraph, then this panic fired off when trying to extract the cyclic egraph (interestingly it *always* chooses the tombstone if given a choice between it and the cycle, even if it's given an absurdly high cost, I assume that's by design though). Is there any way to actually extract cyclic terms from the egraph or plans to implement that (or pointers on implementing it myself)?

Here's the full egraph:
```
0: [UInt(10)]
1: [UInt(1)]
2: [Other("phi", [0, 3])]
3: [Add([2, 13]), Add([13, 2]), Sub([2, 1])]
5: [UInt(0)]
6: [Eq([2, 5])]
7: [Break]
8: [OpaqueFunc(Func { id: 0 })]
9: [Unit]
10: [Apply([8, 9])]
11: [If([6, 7, 10])]
12: [Loop(11)]
13: [Mul([1, 16]), Mul([16, 1]), Neg(1)]
16: [Int(-1)]
```

**Reply by mwillsey** [reply]
It will always choose the tombstone because a cyclic term has "infinite" cost from the perspective of our bottom-up extractor. There is no way to coerce our extractor to give you a cyclic term. The original PEG paper uses ILP to extract cyclic terms from their e-graphs, and the same could work here. I have no doubt that a simpler approach might work as well, but I haven't done. If you come up with something (and you'd need a way to represent cyclic terms as well), I'd love to hear about it!

**Reply by Kixiron** [reply]
So extracting cyclic terms would require writing my own extraction infrastructure? And do you have a link to the PEG paper you're referencing?

**Reply by mwillsey** [reply]
Yes, you'll need to write your own. Here is the paper: https://dl.acm.org/doi/pdf/10.1145/1480881.1480915

**Comment by jameysharp** [comment, marked as answer]
I'm grateful that you mentioned RVSDG here, @Kixiron. Based on your suggestion I read those papers this month, especially ["Perfect Reconstructability of Control Flow from Demand Dependence Graphs"](https://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.687.6305), which convinced me that it's a good model for arbitrary control flow in an e-graph. Re-reading my own blog post from five years ago (["Search-based compiler code generation"](https://jamey.thesharps.us/2017/06/19/search-based-compiler-code-generation/)), I'd apparently discovered the "Perfect Reconstructability" paper then, but by the time I learned about equality saturation I'd forgotten all about it.

And now I have a working prototype of e-graphs for RVSDG, based on egg. Every valid RVSDG is acyclic, so avoids the extraction difficulties discussed here, but you can still round-trip even irreducible control flow through it. I'm very pleased both with egg and with how this experiment has turned out. And yes, to the original question, loops seem to work fine in this model.

https://github.com/jameysharp/optir/

**Reply by mwillsey** [reply]
Hello Jamey!

This is very cool work! I shared this with the other folks in the egg research group, and we all agree that this is a great response to the very common question of "what about loops/control flow?". Keep us posted if you end up using this in a larger setting!
